keps/sig-node/3085-pod-conditions-for-starting-completition-of-sandbox-creation/README.md

```diff
@@ -0,0 +1,1484 @@
+- Look-up pod fields from API server (associated with a pod's trace) like
+labels/annotations/storage classes of PVCs referred by the pod/runtimeclass/etc.
+that is of interest to cluster operators and their users for classifying and
+aggregating the metrics.
+- Look-up a pod's Scheduled condition fields to determine the beginning of pod
+sandbox creation.
+
+Since the lookup of the pod fields and existing conditions is necessary for SLIs
+around pod sandbox creation latency, surfacing the `SandboxReady` condition in
+pod status will allow a metric collection service to directly access the
+relevant data without requiring the ability to collect and parse OpenTelemetry
+traces. As mentioned in the User Stories, popular community managed services
+like Kube State Metrics can consume the `SandboxReady` condition with a trivial
+set of changes. Enhancing them to collect and parse OpenTelemetry traces with no
+sampling and mapping the data to associated data from API server data will be
+complex from an engineering and operational perspective.
+
+For controllers using the pod sandbox conditions to determine reconciliation
+strategy, access to the pod is typically necessary while collecting and parsing
+traces would be unusual.
+
+### Have CSI/CNI/CRI plugins mark their start and completion timestamps while setting up their respective portions for a pod
+
+Each infrastructural plugin that Kubelet calls out to (in the process of setting
+up a pod sandbox) can mark start and completion timestamps on the pod as
+conditions. This approach would be similar to how readiness gates work today.
+However, CSI and CRI plugins will need to be englightened about fields in a pod
+(like status conditions) and setup a client to the API server (to update the
+conditions) which they may not implement to stay orchestrator agnostic.
+
+### Use a dedicated service between Kubelet and CRI runtime to mark sandbox ready condition on a pod
+
+An on-host binary that runs as a service and proxies CRI API calls between the
+CRI runtime and Kubelet can intercept the successful creation of a pod sandbox
+in response to CRI `RunPodSandbox`. Next, using an API server client, the binary
+can mark extended conditions on a pod to indicate state of sandbox creation.
+While this approach works, without requiring any additional changes to Kubelet,
+it had a couple of disadvantages: First, this approach requires configuration
+and management of a separate proxy binary between Kubelet and CRI runtime in the
+cluster nodes. Second, the proxy binary will need to replicate the logic in
+Kubelet status manager to efficiently interact with the API server (as well as
+cache the status and retry in case of API server outages) regarding updates to
+pod sandbox status. Therefore isolating the logic around pod sandbox conditions
+to a separate binary intercepting API calls between kubelet and the CRI runtime
+is not preferred.
```

Comment:
I was imagining that the kubelet itself could report the extended conditions, no need for a separate proxy.

So maybe this is 2 alternatives:
1. Minting new extended conditions and teaching the kubelet to apply these. The difference between this and the KEP
   proposal is that there is no API change, because extended conditions are not part of the Kubernetes API (whereas normal conditions are).
1. Same, but doing it with a CRI proxy and no change to kubelet code.

Of those, I was proposing adding approach #<span>1</span> to the alternatives list.

Reply:
Those are good questions, but it'd need someone who knows more of the (SIG) Architecture side of things to answer them.

I can confirm that the list in https://kubernetes.io/docs/concepts/workloads/pods/pod-lifecycle/#pod-conditions would need to change if we update `…/k8s.io/api/core/v1/types.go`